lib/Serialization/Deserialization.cpp

```diff
@@ -166,10 +167,27 @@ static void skipRecord(llvm::BitstreamCursor &cursor, unsigned recordKind) {
 }
 
 void ModuleFile::fatal(llvm::Error error) const {
+  Core->fatal(diagnoseFatal(std::move(error)));
+}
+
+llvm::Error ModuleFile::diagnoseFatal(llvm::Error error) const {
   if (FileContext)
     getContext().Diags.diagnose(SourceLoc(), diag::serialization_fatal,
                                 Core->Name);
-  Core->fatal(std::move(error));
+  std::string msg;
+  {
+    llvm::raw_string_ostream os(msg);
+    Core->outputDiagnosticInfo(os);
+  }
+  msg += ": " + toString(std::move(error));
+  llvm::dbgs() << "*** DESERIALIZATION FAILURE ***\n";
+  llvm::dbgs() << "*** If any module named here was modified in the SDK, please delete the ***\n";
+  llvm::dbgs() << "*** new swiftmodule files from the SDK and keep only swiftinterfaces.   ***\n";
+  outputDiagnosticInfo(llvm::dbgs());
+  llvm::dbgs() << '\n' << msg << '\n';
+  assert(false && "fatal deserialization failure");
+  return llvm::make_error<llvm::StringError>(msg,
+                                             llvm::inconvertibleErrorCode());
```

Comment:
Should we have a specific error object for these fatal failures? 

We have existing logic that expects any error to be non-fatal and keep going. For example, if one field fails to load we keep reading the other ones. Even if compilation fails later because of the diagnostic, that pre-existing logic will silently ignore the error and keep going to read more. We may need to update all these call sites to prioritize passing up fatal error objects when it's detected.

Reply:
> For example, if one field fails to load we keep reading the other ones.

My first reaction was that that may be okay, but do you think we could end up in an infinite loop when continuing after a fatal error?

Reply:
There's two problems here, first there's a security issue with reading and acting on corrupted memory. The second is not so much an infinite loop, but we may very well see memory access errors. When the compiler reads from a corrupted swiftmodule, it's either rejected by a fatal error on the format, or I've often seen it break at reading an invalid index value and using it as an offset to read something from the swiftmodule. This can lead to more reading from invalid memory or memory access violation errors.

Reply:
Are you suggesting to convert the places that currently return an empty array in the case of failure to return Expected so we can bubble up the failure and abort deserialization?

Reply:
That would be all occurrences of `consumeError` that are not followed by returning a new error.

Reply:
So.. if we wanted to do this we would need to (as you previously suggested) introduce a new kind of error so we can keep the original behavior in places like
```
      if (!pattern) {
        // Silently drop the pattern...
        llvm::consumeError(pattern.takeError());
        // ...but continue to read any further patterns we're expecting.
        continue;
      }
```
and only surface our new fatal error while ignoring all others.

Reply:
Yes exactly this kind of use case. We'd want to consume only the non-fatal errors and pass up the fatal ones immediately.

Reply:
I updated the patch to create a new kind DeserializationError. I haven't passed it through to `ModuleFile::loadAllConformances()` and similar functions, but I can if need be.